packages/stdlib/src/ring1/async.js

```diff
@@ -0,0 +1,49 @@
+// @ts-check
+import { makePromise } from '../ring0/main';
+
+/**
+ * Resolve promise with value when body returns falsy.
+ *
+ * @template T
+ * @param {() => T | Promise<T>} body perform side-effects, and return truthy if
+ * we should run again
+ * @returns {Promise<void>}
+ */
+const asyncDoWhile = body => {
+  const loop = async (resolve, reject) => {
+    const doContinue = await body();
+    if (!doContinue) {
+      // Resolve the outermost promise.
+      resolve(undefined);
+      return;
+    }
+
+    // Do the loop again.  We are careful not to await so that we don't create a
+    // promise chain.
+    loop(resolve, reject).catch(reject);
+  };
+
+  return makePromise((resolve, reject) => loop(resolve, reject).catch(reject));
```

Comment:
Believe it or not, the `.catch(reject)` here is redundant. However, it is clearer with it. If you do remove it, please insert a comment explaining why something like that isn't needed.

Reply:
Ah, good catch.  I will remove and put a comment because I understand why.

Reply:
Oh, maybe I don't.  I'll leave it in place.